File: objectiveC/SKAConversionValues.m
//
//  SKAConverionValues.m
//  DataseatTrackingTester
//
//  Created by Paul Hayton on 01/02/2021.
//  Copyright © 2021 Paul Hayton. All rights reserved.
//

#import <Foundation/Foundation.h>

#import "SKAConversionValues.h"

@implementation SKAConversionValues

+ (void)registerAppOpen {
    
    if (@available(iOS 14.0, *)) {
        NSInteger currentValue = [[NSUserDefaults standardUserDefaults] integerForKey:@"SKAConversionValue"];
        if (currentValue == 0) {
            NSCalendar *cal = [NSCalendar currentCalendar];
            [cal setTimeZone:[NSTimeZone timeZoneWithAbbreviation:@"UTC"]];
            NSDateComponents *dateComponents = [cal components:NSCalendarUnitWeekday fromDate:[NSDate date]];
            long weekday = [dateComponents weekday];
            [SKAdNetwork updateConversionValue:weekday & 0x07];
            // Store this value so we can update it.
            [[NSUserDefaults standardUserDefaults] setObject:[NSNumber numberWithLong:weekday] forKey:@"SKAConversionValue"];
            [[NSUserDefaults standardUserDefaults] synchronize];
        }
    }
}

+ (void)updateConversionValue:(int)value {
    if (@available(iOS 14.0, *)) {
        NSInteger upper3Bits = (value & 0x07) << 3;
        NSInteger currentValue = [[NSUserDefaults standardUserDefaults] integerForKey:@"SKAConversionValue"];
        NSInteger lower3Bits = currentValue & 0x07;
        
        NSInteger newValue = upper3Bits | lower3Bits;
        if (newValue > currentValue) {
            [SKAdNetwork updateConversionValue:newValue];
            // Store this value so we can update it.
            [[NSUserDefaults standardUserDefaults] setObject:[NSNumber numberWithLong:newValue] forKey:@"SKAConversionValue"];
            [[NSUserDefaults standardUserDefaults] synchronize];
        }
    }
}

@end

File: swift/SKAConversionValues.swift
//
//  SKAConversionValues.swift
//  flappyjoust
//
//  Created by Paul Hayton on 01/02/2021.
//

import Foundation

class SKAConversionValues: NSObject {

    static func registerAppOpen() {
        if #available(iOS 14.0, *) {
            let defaults = UserDefaults.standard
            let currentValue = defaults.integer(forKey: "SKAConversionValue")
            if (currentValue == 0) {
                var cal = Calendar(identifier: .gregorian)
                cal.timeZone = TimeZone(identifier: "UTC")!
                let dateComponents = cal.dateComponents([.weekday], from: Date())
                let weekDay = dateComponents.weekday!
                SKAdNetwork.updateConversionValue(weekDay & 0x07)
                defaults.setValue(weekDay, forKey: "SKAConversionValue")
                defaults.synchronize()
            }
        }
    }
    
    static func updateConversionValue(value : Int) {
        if #available(iOS 14.0, *) {
            let upper3bits = (value & 0x07) << 3
            let defaults = UserDefaults.standard
            let currentValue = defaults.integer(forKey: "SKAConversionValue")
            let lower3bits = currentValue & 0x07
            
            let newValue = upper3bits | lower3bits
            if (newValue > currentValue) {
                SKAdNetwork.updateConversionValue(newValue)
                defaults.setValue(newValue, forKey: "SKAConversionValue")
                defaults.synchronize()
            }
        }
    }
}
